### PR TITLE
[`Add Mixtral`] Adds support for the Mixtral MoE

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?

Adds the latest MoE model from mistral AI

### CODE REVIEW

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 1]
```diff
@@ -0,0 +1,1451 @@
...
+    def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        """ """
+        batch_size, sequence_length, hidden_dim = hidden_states.shape
+        hidden_states = hidden_states.view(-1, hidden_dim)
+        # router_logits: (batch * sequence_length, n_experts)
+        router_logits = self.gate(hidden_states)
+
+        routing_weights, selected_experts = torch.topk(router_logits, self.top_k, dim=-1)
+        routing_weights = routing_weights.softmax(dim=-1)
+
```
Comment (casper-hansen) [inline src/transformers/models/mixtral/modeling_mixtral.py:711]: Following the PR from vLLM from the Mistral team, the order of operations is as follows:
gate -> softmax -> topk -> sum
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/mixtral.py#L339-L344

Comment (152334H) [reply src/transformers/models/mixtral/modeling_mixtral.py:711]: it should be noted that `gate -> softmax -> topk -> sum` is arithmetically equivalent to `gate -> topk -> softmax`, ignoring numerical precision errors

Comment (casper-hansen) [reply src/transformers/models/mixtral/modeling_mixtral.py:711]: Unfortunately, it's not precisely equivalent which is why it must be fixed before merging it to avoid higher loss when training.

Comment (ArthurZucker) [reply src/transformers/models/mixtral/modeling_mixtral.py:711]: We tried both, had the same results but will update for sure 🤗

File: docs/source/en/model_doc/mixtral.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+## Overview
+
+Mixtral-8x7B is Mistral AI's second Large Language Model (LLM). 
+
+The Mixtral model was proposed in the [Mistral AI team](https://mistral.ai/).
+
+
+The abstract from the paper is the following:
+
+*<INSERT PAPER ABSTRACT HERE>*
```
Comment (LysandreJik) [inline docs/source/en/model_doc/mixtral.md:28]:
```suggestion
The Mixtral model was proposed in the by the [Mistral AI](https://mistral.ai/) team.

It was introduced in the [Mixtral of Experts blogpost](https://mistral.ai/news/mixtral-of-experts/) with the following introduction:

*Today, the team is proud to release Mixtral 8x7B, a high-quality sparse mixture of experts models (SMoE) with open weights. Licensed under Apache 2.0. Mixtral outperforms Llama 2 70B on most benchmarks with 6x faster inference. It is the strongest open-weight model with a permissive license and the best model overall regarding cost/performance trade-offs. In particular, it matches or outperforms GPT3.5 on most standard benchmarks.*
```

File: docs/source/en/model_doc/mixtral.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+Tips:
+
+
+- The model needs to be converted using the [conversion script](https://github.com/huggingface/transformers/blob/main/src/transformers/models/mixtral/convert_mixtral_weights_to_hf.py).
```
Comment (LysandreJik) [inline docs/source/en/model_doc/mixtral.md:33]: Maybe not relevant once the weights are pushed to the Hub

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 4]
```diff
@@ -0,0 +1,1451 @@
...
+        attn_output = self.o_proj(attn_output)
+
+        if not output_attentions:
+            attn_weights = None
+
+        return attn_output, attn_weights, past_key_value
+
+
+# Copied from transformers.models.mistral.modeling_mistral.MistralFlashAttention2 with Mistral->Mixtral
+class MixtralFlashAttention2(MixtralAttention):
```
Comment (LysandreJik) [inline src/transformers/models/mixtral/modeling_mixtral.py:371]: :fire:

File: src/transformers/models/mixtral/modeling_mixtral.py [hunk 5]
```diff
@@ -0,0 +1,1451 @@
...
+
+            # However `index_add_` only support torch tensors for indexing so we'll use
+            # the `top_x` tensor here.
+            final_hidden_states.index_add_(0, top_x, current_hidden_states.to(hidden_states.dtype))
+        final_hidden_states = final_hidden_states.reshape(batch_size, sequence_length, hidden_dim)
+        return final_hidden_states, router_logits
+
+
+class MixtralDecoderLayer(nn.Module):
+    # Ignore copy
```
Comment (LysandreJik) [inline src/transformers/models/mixtral/modeling_mixtral.py:746]:
```suggestion
```

File: src/transformers/models/mixtral/convert_mixtral_weights_to_hf.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--input_dir",
+        help="Location of Mistral weights, which contains tokenizer.model and model folders",
+        required=True,
+    )
+    parser.add_argument(
+        "--model_size",
+        choices=["7B"],
+        help="'f' models correspond to the finetuned versions, and are specific to the Mistral2 official release. For more details on Mistral2, checkout the original repo: https://huggingface.co/meta-mistral",
```
[the remark this answers is not in the document]
Comment (ArthurZucker) [reply src/transformers/models/mixtral/convert_mixtral_weights_to_hf.py:229]: Thanks updating 😉

**Review by LysandreJik — APPROVED** [review]
Looks awesome! Thanks for the integration @ArthurZucker @younesbelkada, merge once you're ready and CI is green. 

A few documentation things that we can improve on, but let's do that after it has landed in the lib.

File: src/transformers/models/mixtral/convert_mixtral_weights_to_hf.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+            state_dict[expert_key + ".weight"] = expert_block.clone()
+
+        experts_w2 = [
+            w2[ffn_dim * expert_idx : ffn_dim * (expert_idx + 1), :].contiguous().clone()
+            for expert_idx in range(num_local_experts)
+        ]
+
+        for idx, expert_block in enumerate(experts_w2):
+            expert_key = f"model.layers.{layer_i}.block_sparse_moe.experts.{idx}.w2"
+            state_dict[expert_key + ".weight"] = expert_block.T.clone()
```
Comment (younesbelkada) [inline src/transformers/models/mixtral/convert_mixtral_weights_to_hf.py:166]:
```suggestion
            state_dict[expert_key + ".weight"] = expert_block.T.clone().contiguous()
```

**Review by younesbelkada — COMMENTED** [review]
A small nit on the conversion script!

File: src/transformers/models/mixtral/modeling_mixtral.py (hunk 1 above)
Comment (casper-hansen) [reply src/transformers/models/mixtral/modeling_mixtral.py:711]: Good to hear @ArthurZucker. Really looking forward to trying this out!

**Comment by ArthurZucker** [comment]
The auxiliary loss can be computed with `output_router_logits = True` automatically, other losses like `z_loss` can be imported from Switch transformers modeling code! Custom losses should be able to use the router_logits returned by the model
